### PR TITLE
Various improvements

### PR DESCRIPTION
**Pull request by francois-berder**
* Fix build failure (obj directory was not created and SDL link errors)
  * Improve Makefile to support parallel builds
  * Fix warnings
  * Fix memory leaks in main
  * Better handling cartridge load failure

### CODE REVIEW

**Review by jc-SpaceXp — APPROVED** [review]
Approved.

**Comment by jc-SpaceXp** [comment]
I didn't mean to approve all 12 commits initially, just one I was looking at.

**Comment by jc-SpaceXp** [comment]
Many of these changes overlap with the cleanup branch that I've been working on recently i.e. malloc checking and SDL_Delay(). I will look over these changes over the next few days and merge appropriately.

**Review by jc-SpaceXp — CHANGES_REQUESTED** [review]
Can you please change the branch from master to cleanup. I understand you made the changes before I made the branch. Also can the missing void commit and the malloc checks for cpu and ppu structs be removed? They've been fixed in the cleanup branch a while ago and will create a small merge conflict.

**Comment by francois-berder** [comment]
Hello,
In that case, I will close this pull request and open one later. Let me know if you merge cleanup branch to master.
François

**Comment by jc-SpaceXp** [comment]
Hi,
I've just merged my cleanup branch to master.

**Comment by francois-berder** [comment]
Hi, I've just rebased my patches on top of the master branch.

**Comment by jc-SpaceXp** [comment]
Hi, I'm going to look through these changes soon!

File: makefile [hunk 1]
```diff
@@ -1,13 +1,29 @@
...
+ifneq ($(CONFIG),release)
+$(error "CONFIG must be set to release or debug")
+endif
+endif
+
+ifeq ($(CONFIG),debug)
+CFLAGS += -g
+endif
+ifeq ($(CONFIG),release)
+CFLAGS += -O2
```
Comment (jc-SpaceXp) [inline makefile:19]: O1, O2 and O3 break emulator

File: makefile [hunk 2]
```diff
@@ -1,13 +1,29 @@
 CC := gcc
-CFLAGS := -Wall -Wextra -std=c99 -g
+CFLAGS := -Wall -Wextra -std=c99
 CFLAGS += $(shell pkg-config --cflags sdl2)
 LDFLAGS := $(shell pkg-config --libs sdl2)
 DEPFLAGS = -MMD -MP -MF $(@:$(OBJDIR)/%.o=$(DEPDIR)/%.d)
 
+CONFIG ?= debug
```
Comment (jc-SpaceXp) [inline makefile:7]: DEBUG ?= 0

Then running: make DEBUG=0 seems better to me
Leads to a simpler if-else chain in the makefile:
```make
DEBUG ?= 0
ifeq ($(DEBUG), 1)
        CFLAGS += -g -D__DEBUG__
        CONFIG = debug
else
        CFLAGS += -DN__DEBUG__
        CONFIG = release
endif
```

File: makefile [hunk 3]
```diff
@@ -1,13 +1,29 @@
...
+
+ifneq ($(CONFIG),debug)
+ifneq ($(CONFIG),release)
+$(error "CONFIG must be set to release or debug")
+endif
+endif
+
+ifeq ($(CONFIG),debug)
+CFLAGS += -g
```
Comment (jc-SpaceXp) [inline makefile:16]: CFLAGS += -g -D__DEBUG__

File: makefile (hunk 1 above)
Comment (jc-SpaceXp) [inline makefile:19]: CFLAGS += -DN__DEBUG__

**Pull request merged**